### PR TITLE
Change Type() signature to return interface and error

### PR DESCRIPTION
**Pull request by nodivbyzero**
This PR changes TypeMatcher's Type() signature to return interface and error.

This change is needed to address an [issue](https://github.com/rstudio/package-manager/issues/7714 ) where the offline node is receiving messages for queue work (work ready, work complete), but it is not configured to handle those message types. This error is innocuous, but could be alarming to customers.

### CODE REVIEW

File: pkg/rsnotify/listener/genericmatcher_test.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	type TestNotification struct{}
+	m := &GenericMatcher{
+		field: "test",
+		types: make(map[uint8]interface{}),
+	}
+	m.Register(testNotificationType, TestNotification{})
+	c.Assert(m.types[testNotificationType], check.NotNil)
+
+	t, _ := m.Type(testNotificationType)
+	c.Assert(t, check.NotNil)
```
Comment (jonyoder) [inline pkg/rsnotify/listener/genericmatcher_test.go:58]: Can we check the `err` being returned to prove it is nil?

Also, can we add a test where we pass an unknown type and verify that it returns `nil` both for the type and the error?

File: pkg/rsnotify/listener/listener_test.go [hunk 2]
```diff
@@ -3,13 +3,9 @@ package listener
...
-
 	"gopkg.in/check.v1"
 )
 
-type NotifySuite struct{}
-
-var _ = check.Suite(&NotifySuite{})
+type ListenerSuite struct{}
 
-func TestPackage(t *testing.T) { check.TestingT(t) }
```
Comment (jonyoder) [inline pkg/rsnotify/listener/listener_test.go:15]: Removing this is causing the build to fail. You need one call like this per package. You can either revert this one or move it to `genericmatcher_test.go`.

File: pkg/rsnotify/listeners/postgrespgx/pgxlistener.go [hunk 3]
```diff
@@ -224,13 +224,17 @@ func (l *PgxListener) notify(n *pgconn.Notification, errs chan error, items chan
 		errs <- fmt.Errorf("error unmarshalling message data type: %s", err)
 		return
 	}
-	if l.matcher.Type(dataType) == nil {
+	t, err := l.matcher.Type(dataType)
+	if err != nil {
+		return
+	}
+	if t == nil {
```
Comment (jonyoder) [inline pkg/rsnotify/listeners/postgrespgx/pgxlistener.go:231]: If you want to be able to ignore errors, you'll need to remove this block. I think we probably want this behavior in the generic matcher. Then, using the generic matcher will return this error when no type is found, but we can create a custom matcher for offline mode that returns no error.

File: pkg/rsnotify/listeners/postgrespq/pqlistener.go [hunk 4]
```diff
@@ -183,13 +183,17 @@ func (l *PqListener) notify(n *pq.Notification, errs chan error, items chan list
 		errs <- fmt.Errorf("error unmarshalling message data type: %s", err)
 		return
 	}
-	if l.matcher.Type(dataType) == nil {
+	t, err := l.matcher.Type(dataType)
+	if err != nil {
+		return
+	}
+	if t == nil {
```
Comment (jonyoder) [inline pkg/rsnotify/listeners/postgrespq/pqlistener.go:190]: See comment on pgxlistener.go.

**Review by jonyoder — COMMENTED** [review]
This looks like it's almost ready. One concern about error handling, and I noticed a test-wiring issue.

File: pkg/rsnotify/listeners/postgrespgx/pgxlistener.go (hunk 3 above)
Comment (nodivbyzero) [reply pkg/rsnotify/listeners/postgrespgx/pgxlistener.go:231]: oops. 
I did the opposite: genericMatcher return no error (no changes in the GenericMatcher) and OfflineMatcher return `MissingType` error.

What kind of interface should OfflineMatcher return in case of missing type?

Comment (jonyoder) [reply pkg/rsnotify/listeners/postgrespgx/pgxlistener.go:231]: I was thinking that the generic matcher would return the error on a missing type, while the offline matcher would simply return a nil error. Then, in the listener implementations, if we get back a nil value and no error, we would simply return with no error.

File: pkg/rsnotify/listener/genericmatcher.go [hunk 5]
```diff
@@ -12,7 +18,10 @@ func (m *GenericMatcher) Field() string {
 }
 
 func (m *GenericMatcher) Type(notifyType uint8) (interface{}, error) {
-	return m.types[notifyType], nil
+	if t, ok := m.types[notifyType]; ok {
+		return t, nil
+	}
+	return nil, MissingTypeError
```
Comment (jonyoder) [inline pkg/rsnotify/listener/genericmatcher.go:24]: I think this is a good case for error wrapping:

```go
return fmt.Errorf("no matcher type found for %d: %w", notifyType, MissingTypeError)
```

File: pkg/rsnotify/listeners/postgrespgx/pgxlistener.go [hunk 6]
```diff
@@ -226,10 +226,10 @@ func (l *PgxListener) notify(n *pgconn.Notification, errs chan error, items chan
 	}
 	t, err := l.matcher.Type(dataType)
 	if err != nil {
+		errs <- fmt.Errorf("no matcher type found for %d", dataType)
```
Comment (jonyoder) [inline pkg/rsnotify/listeners/postgrespgx/pgxlistener.go:229]: Here, and also in the pqlistener, if we switch to error wrapping, we can simply pass the error here:
```
errs <- err
```

File: pkg/rsqueue/groups/grouprunner.go [hunk 7]
```diff
@@ -41,7 +42,10 @@ func (m *GenericMatcher) Field() string {
 }
 
 func (m *GenericMatcher) Type(workType uint64) (interface{}, error) {
-	return m.types[workType], nil
+	if t, ok := m.types[workType]; ok {
+		return t, nil
+	}
+	return nil, MissingTypeError
```
Comment (jonyoder) [inline pkg/rsqueue/groups/grouprunner.go:48]: We can wrap this one, too, and then simply pass the error below as suggested above.

**Review by jonyoder — COMMENTED** [review]
Just one more suggestion about wrapping errors to make them easier to pass later. After that, I think this is good to merge.

File: pkg/rsnotify/listener/genericmatcher.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
+package listener
+
```
Comment (jonyoder) [inline pkg/rsnotify/listener/genericmatcher.go:2]: CI is failing since there is no copyright notice on line 3 here.

**Review by jonyoder — APPROVED** [review]
When the copyright notice is fixed and CI passes, let's merge this!

**Comment by jonyoder** [comment]
FYI, you can catch these issues before pushing if you run:
```
just vet
just lint
```
